### ISSUE TITLE
Support values files for helm operator

### ISSUE THREAD

**Issue by the-fine** [weaveworks/flux#1349]
Currently we need to specify the values in the FluxHelmRelease spec, it would be great if we could specidy the value files locationa inside the git repo

**Comment by squaremo** [comment]
I can see how it would be a bit more readable than values inlined in the FluxHelmRelease. Are there other reasons too prefer doing things this way?

Which git repo do you have in mind? There is the one that has chart sources in it, that you tell `helm-operator` about; and there's the one with FluxHelmReleases (and other YAMLs) in it, which you tell `fluxd` about.

I guess since the values relate to the FluxHelmRelease resources, they should be alongside those files in the latter git repo, the one provided to fluxd. But helm-operator cannot see this repo, in general, and won't have access to those files. :thinking: That means either fluxd has to treat FluxHelmRelease YAMLs as a special case when applying them; or, some other completely different design. I quite like the limited role of fluxd ("apply what's in git"); I am less confident about how helm-operator works.

**Comment by wstrange** [comment]
Using a an external values.yaml helps to minimize the difference between deploying with flux vs. deploying manually with `helm install`. The  later is useful in development or for doing adhoc testing.

**Comment by sfitts** [comment]
For us another reason for this is DRY-ness (at least for us).

In our environment we have to support target clusters from multiple providers (aws, alicloud, azure) so for some of the components we install we have provider specific values values that get "overlayed" into the helm command when we deploy.  We also have values that are associated with the target environment vs. those that are standard for all of our deployments.  So it is pretty typical that we'll be including 3 (or more) values files on a single helm command line (ordering from most general to most specific so overrides work as expected).

If I understand correctly, with Flux today, each of these would turn into a unique FHR, meaning we'd have quite a few for each component and end up repeating the same configuration multiple times.  This seems like it will cause maintenance issues.

We could probably figure out a way to generate many of these and propagate the changes into all the places needed, but that works against the whole "version everything" approach since versioning generated artifacts is generally problematic.

**Comment by squaremo** [comment]
A proposal over at #1468 is to let you refer to secrets from the FHR:

```
...
spec:
  valueSecrets:
  - name: clusterSpecifics
  - name: furtherSpecifics
  values:
    key: value
    as: usual
```

The mode of use would be:

 1. put a secret (possibly sealed) with an entry `values.yaml` in the git repo, or create it directly in the cluster
 2. refer to it in the FluxHelmRelease

I think this covers the DRY requirement. It is not quite as convenient as a flat-out `values.yaml` file, somewhere, but it does avoid the "wrong repo" problem I described above. It also covers #1172, I believe.

**Comment by sfitts** [comment]
@squaremo thanks for the pointer.

Not sure where the appropriate place to comment on this is, so I'll keep going here (don't want to clutter up the PR since I'm very much a newbie here).

If I understand #1468 correctly it relies on having a config map pre-mounted into the helm operator pod.  This seems reasonable for expressing values that are:

a) Cluster specific
b) Generally applicable across multiple charts (aka fairly global)
c) Not sensitive

I have some concerns about using this mechanism to manage the values files we have today, but if I'm understanding your suggested changes correctly those may be addressed.

In your comments you were suggesting the switch to/addition of secrets.  Was your picture that these would still be pre-mounted into the helm pod (so they would manifest as files in a well known location on disk) or were you suggesting that the operator directly read these secrets to generate the final values?

If it was the latter than this seems much better suited to our needs.  For every chart we install we will have the possibility for multiple values "files" (collections of values that are applied together) and managing all of these via a single config map (while making sure that there are no name collisions/etc...) would be untenable.  Also, these values will include sensitive information, so some of them can't be stored in a CM (or in clear text in Git either).

Your example above also implies the ability to specify multiple such secrets, with potentially overlapping value keys allowing us to control the application order and get the same end result we do today.

Is this correct or am I missing something?

One other question -- how do you envision that the secrets themselves would be deployed?  Is that something that would be handled by base Flux (I assume yes, but wanted to be sure)?  If so, is there a mechanism that would allow the helm operator to know when those have changed so it can perform an "upgrade" to the impacted release/releases?

**Comment by squaremo** [comment]
@sfitts 

> Was your picture that these would still be pre-mounted into the helm pod (so they would manifest as files in a well known location on disk) or were you suggesting that the operator directly read these secrets to generate the final values?

Yes, this latter -- pretty close to the original design in that PR, but looking up secrets from K8s rather than files on disk.

Good to have confirmation that this is going to cover this use case for you!

> One other question -- how do you envision that the secrets themselves would be deployed?

Glib answer: however you want :-) But, yes, using fluxd when it's sensible (e.g., in conjunction with SealedSecrets).

With respect to responding to changes in the secrets, the operator does a "full reconciliation" periodically (on the order of low numbers of minutes); initially, I think we'll probably let that fallback account for changing secrets. It's certainly possible to install watches to be more responsive if that turns out to be inadequate.

**Comment by squaremo** [comment]
#1468 is merged -- if you have a chance, it's worth trying it out to see if it covers your use cases. It may be in a coming v0.3.1, and it'll be ported to the "beta" code in #1382.

**Comment by corinz** [comment]
@the-fine @squaremo Can we reopen this? The link between my Flux HelmRelease and my Helm Chart is a GitRepository that has a path to the Chart.yaml. If I cannot include the path to values.yaml as well, I am forced to specify them statically from the command line for injection into my Flux HelmRelease. This is possible but requires cli and creates static values hardcoded to the Flux HelmRelease. My chart will be GitOps enabled, but my values will need to be adjusted in separate modifications to the HelmRelease. This would double the values line count and create a disjointed system config.

Can you help? This is a make or break for me unfortunately. Do you have any other suggestions.

I would appreciate any feedback! Thank you!

```yaml
---
apiVersion: helm.toolkit.fluxcd.io/v2beta1
kind: HelmRelease
metadata:
  name: monitoring
  namespace: flux-system
spec:
  chart:
    spec:
      chart: ./ala-monitoring
      sourceRef:
        kind: GitRepository
        name: flux-system
  install:
    createNamespace: true
  interval: 5s
  targetNamespace: monitoring
# Hard-coding values here is an odd quirk in an otherwise seamless workflow
  values:
    grafana:
      enabled: true
```

**Comment by kingdonb** [comment]
@corinz I am not sure about the details of this issue, but it looks like you are using Flux v2, this issue was reported against Flux v1. If it still represents a problem in Flux v2, it will have to be addressed there separately, but as I said I'm not sure about the details myself, I don't have a clear understanding of your issue, but maybe this issue was addressed already?

* https://fluxcd.io/docs/guides/helmreleases/#refer-to-values-inside-the-chart

If this documentation helps you solve your issue, great!

If not, can you open a new issue on the Helm Controller or find a relevant issue that is open against fluxcd/flux2 repository, and report there (where everyone who works on Flux now will see it?) Very few people still visit fluxcd/flux, for legacy Flux v1.

Thanks for using Flux!
